### ISSUE TITLE
False Negative: Syntactic variants of redundant casts are no longer recognized.

### ISSUE THREAD

**Issue by Carlson-JLQ** [pmd/pmd#6538]
# False Negative: Syntactic variants of redundant casts are no longer recognized.

Version
7.22.0

## Checker
- Checker id: `category/java/codestyle.xml/UnnecessaryCast`
- Checker description: This rule detects unnecessary cast expressions where the cast is redundant because the operand's type already satisfies the context's requirements or the cast does not affect the semantics.

## Description of the false negative
`diff/output` marks 1 false negatives for `category/java/codestyle.xml/UnnecessaryCast`. Across these samples, the benchmark is still exercising the same underlying bug pattern, but the implementation appears to lose track of it once the source deviates from one canonical shape. The misses read more like matching gaps than deliberate exclusions.

### `PosCase4.java`
Source: `output/sonarqube/RedundantTypeCastCheck/src/main/java/scensct/core/pos/PosCase4.java`

The sample keeps the benchmark's target pattern but expresses it in a slightly non-canonical source form. The risky pattern is still present in the code, so the absence of a report points to a matching gap rather than a benign variant. That is also how the benchmark classifies the sample: Cast of null literal used as receiver (non-getClass method) and not in argument list should be flagged as unnecessary.

```java
// Cast of null literal used as receiver (non-getClass method) and not in argument list should be flagged as unnecessary.
package scensct.core.pos;

public class PosCase4 {
    public static void main(String[] args) {
        // Cast of null literal used as receiver for toString() method
        String str = ((String) null).toString(); // [REPORTED LINE]
        System.out.println(str);
    }
}
```

## Cause analysis
The rule is documented as follows: This rule detects unnecessary cast expressions where the cast is redundant because the operand's type already satisfies the context's requirements or the cast does not affect the semantics.

 The missed cases suggest that the implementation is narrower than the rule description and too dependent on one preferred AST shape.

Once the same behavior is expressed through a helper, an overload, a modifier such as `abstract` or `native`, a different comment position, or a slightly rearranged control-flow structure, the report disappears even though the benchmark still treats the case as in-scope.

For `category/java/codestyle.xml/UnnecessaryCast`, the practical improvement would be to generalize the match so it follows the underlying behavior rather than one exact spelling of it.

## References
None known.

**Comment by oowekyala** [comment]
The cast is not unnecessary. Calling a method on the null literal produces a type error.
